### PR TITLE
feat(gitlab): Enable retry on failures while fetching secrets

### PR DESCRIPTION
**Pull request by chouetz**
### What does this PR do?
Alternative to #29087
Change the way we store secrets in script. Instead of 
```
export VAR=$(fetch_secret)
```
use
```
VAR=$(fetch secret) || exit $?; export VAR
```
To be able to use the `retry:exit_codes` feature from gitlab, we need to throw the error in gitlab configuration. As a consequence, we need to split assignation and export (cf [this shell check](https://www.shellcheck.net/wiki/SC2155)), otherwise the return value of `fetch_secret` would be ignored

### Motivation
Prevent the failures when emissary container is down before the job starts.

### Possible Drawbacks / Trade-offs
Contrary to the solution proposed on #29087, with this implementation we cannot do:
```
VAR=$(fetch_secret) run_command_needing_VAR
```
And in some cases (regression_detector) we added more storage on variables because it's not possible to do
```
$(fetch_secret) | use_secret_from_stdin
```
neither. However we leverage the `PIPESTATUS` [array](https://unix.stackexchange.com/questions/14270/get-exit-status-of-process-thats-piped-to-another) to send password on stdin when possible (docker).
While the above is the [recommended way](https://datadoghq.atlassian.net/wiki/spaces/SECENG/pages/3826647944/Secure+Storage+and+Handling+of+Secrets#Tips-and-Tricks) we can see in  #29087 it adds complexity to the gitlab configuration and impacts readability.

### Describe how to test/QA your changes
Pipeline is successful

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Review by L3n41c — COMMENTED** [review]
Would it make sense to use `set -e` as prelude of all scripts instead of adding `|| exit $?` at many places ?

It would change the default behavior to exit in case of error by default.

If there are some cases where we want to explicitly ignore errors, we can add `||:` there but I feel that there should be more places where we want to catch errors than places where we want to ignore them.

**Review by L3n41c — APPROVED** [review]

File: .gitlab/container_build/docker_linux.yml [hunk 1]
```diff
@@ -13,8 +13,9 @@
       fi
     - TARGET_TAG=${IMAGE}${ECR_RELEASE_SUFFIX}:v${CI_PIPELINE_ID}-${CI_COMMIT_SHORT_SHA}$TAG_SUFFIX-$ARCH
     # DockerHub login for build to limit rate limit when pulling base images
-    - DOCKER_REGISTRY_LOGIN=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $DOCKER_REGISTRY_LOGIN)
-    - $CI_PROJECT_DIR/tools/ci/fetch_secret.sh $DOCKER_REGISTRY_PWD | docker login --username "$DOCKER_REGISTRY_LOGIN" --password-stdin "$DOCKER_REGISTRY_URL"
+    - DOCKER_LOGIN=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $DOCKER_REGISTRY_LOGIN) || exit $?
+    - DOCKER_PWD=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $DOCKER_REGISTRY_PWD) || exit $?
+    - docker login --username "$DOCKER_LOGIN" --password "$DOCKER_PWD" "$DOCKER_REGISTRY_URL"
```
[the remark this answers is not in the document]
Comment (chouetz) [reply .gitlab/container_build/docker_linux.yml:18]: Yes I agree. The alternative to this (where I can be even more secure) stays in #29087. As I say I the description, this solution is rather complex. We will assess what is the best trade-off

File: .gitlab/e2e_k8s/e2e_k8s.yml [hunk 2]
```diff
@@ -11,16 +11,16 @@
...
+    - DOCKER_REGISTRY_PWD=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $DOCKER_REGISTRY_PWD) || exit $?; export DOCKER_REGISTRY_PWD
+    - DD_API_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $API_KEY_DDDEV) || exit $?; export DD_API_KEY
 
 .k8s-e2e-cws-cspm-init:
   - set +x
-  - export DATADOG_AGENT_SITE=datadoghq.com
-  - export DATADOG_AGENT_API_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $E2E_TESTS_API_KEY)
-  - export DATADOG_AGENT_APP_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $E2E_TESTS_APP_KEY)
-  - export DATADOG_AGENT_RC_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $E2E_TESTS_RC_KEY)
+  - DATADOG_AGENT_SITE=datadoghq.com || exit $?; export DATADOG_AGENT_SITE
```
[the remark this answers is not in the document]
Comment (chouetz) [reply .gitlab/e2e_k8s/e2e_k8s.yml:20]: Nice catch, I fixed several of these because my replace script was too wide and missed this one, thanks

**Review by gjulianm — APPROVED** [review]

File: .gitlab/functional_test/regression_detector.yml [hunk 3]
```diff
@@ -42,12 +42,14 @@ single-machine-performance-regression_detector:
...
     - SMP_ECR_URL=${SMP_ACCOUNT_ID}.dkr.ecr.us-west-2.amazonaws.com
-    - SMP_AGENT_TEAM_ID=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_AGENT_TEAM_ID)
-    - SMP_API=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_API)
-    - aws configure set aws_access_key_id $($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_BOT_ACCESS_KEY_ID) --profile ${AWS_NAMED_PROFILE}
-    - aws configure set aws_secret_access_key $($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_BOT_ACCESS_KEY) --profile ${AWS_NAMED_PROFILE}
+    - SMP_AGENT_TEAM_ID=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_AGENT_TEAM_ID) || exit $?
+    - SMP_API=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_API) || exit $?
+    - SMP_BOT_ID=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_BOT_ACCESS_KEY_ID) || exit $?
+    - SMP_BOT_KEY=$($CI_PROJECT_DIR/tools/ci/fetch_secret.sh $SMP_BOT_ACCESS_KEY) || exit $?
+    - aws configure set aws_access_key_id "$SMP_BOT_ID" --profile ${AWS_NAMED_PROFILE}
```
Comment (chouetz) [inline .gitlab/functional_test/regression_detector.yml:51]: It's not possible here to do
```
command $(subcommand)
```
and get the exit code of the subcommand so I must store in a variable here

**Comment by gjulianm** [comment]
I just noticed that we are missing adding `exit_codes: 42` to the [KMT retry logic](https://github.com/DataDog/datadog-agent/blob/usama.saqib/alien-vms/.gitlab/kernel_matrix_testing/common.yml#L228) too, as it's different from the default one.

**Review by amenasria — APPROVED** [review]

**Review by cmetz100 — APPROVED** [review]

**Pull request merged**